include/sacn/cpp/universe_discovery.h

```diff
@@ -0,0 +1,399 @@
+ * 
+ * // What if your network interfaces change? Update my_netints and call this:
+ * etcpal::Error reset_result = my_universe_discovery.ResetNetworking(my_netints);
+ * // Check reset_result here...
+ * 
+ * // To destroy a universe discovery listener, call this:
+ * my_universe_discovery.Shutdown();
+ * 
+ * // During application shutdown, everything can be cleaned up by calling sacn::Deinit.
+ * sacn::Deinit();
+ * @endcode
+ *
+ * Callback demonstrations:
+ * @code
+ * void MyNotifyHandler::HandleUpdateSource(const etcpal::Uuid& cid, const std::string& name,
+ *                                          const uint16_t* sourced_universes, size_t num_sourced_sources)
+ * {
+ *   std::cout << "Universe discovery: Source " << cid.ToString() << " (name " << name << ") ";
+ *   if(sourced_universes)
+ *   {
+ *     std::cout << "is active on these universes: ";
+ *     for(size_t i = 0; i < num_sourced_universes; ++i)
+ *       std::cout << sourced_universes[i] << " ";
+ *     std::cout << "\n";
+ *   }
+ *   else
+ *   {
+ *     std::cout << "is not active on any universes.\n";
+ *   }
+ * }
+ *
+ * void MyNotifyHandler::HandleSourceExpired(const etcpal::Uuid& cid, const std::string& name)
+ * {
+ *   std::cout << "Universe discovery: Source " << cid.ToString() << " (name " << name << ") has expired.\n";
+ * }
+ *
+ * void MyNotifyHandler::HandleMemoryLimitExceeded()
+ * {
+ *   std::cout << "Universe discovery: Source/universe limit exceeded!\n";
+ * }
+ * @endcode
+ */
+
+class UniverseDiscovery
+{
+public:
+  /** A handle type used by the sACN library to identify Universe Discovery listener instances. */
+  using Handle = sacn_universe_discovery_t;
+  /** An invalid Handle value. */
+  static constexpr Handle kInvalidHandle = SACN_UNIVERSE_DISCOVERY_INVALID;
+
+  /**
+   * @ingroup sacn_universe_discovery_cpp
+   * @brief A base class for a class that receives notification callbacks from a sACN Universe Discovery listener.
+   */
+  class NotifyHandler
+  {
+  public:
+    virtual ~NotifyHandler() = default;
+
+    /**
+     * @brief Notify that a source is new or has changed.
+     *
+     * This passes the source's current universe list, but you will only get this callback when the module detects
+     * that the source is new or the list has somehow changed.
+     *
+     * The list of sourced universes is guaranteed by the protocol to be numerically sorted.
+     *
+     * @param[in] cid The CID of the source.
+     * @param[in] name The UTF-8 name string.
+     * @param[in] sourced_universes Numerically sorted array of the currently sourced universes.  Will be NULL if the
+     * source is not currently transmitting any universes.
+     * @param[in] num_sourced_universe Size of the sourced_universes array.  Will be 0 if the source is not currently
+     * transmitting any universes.
+     */
+    virtual void HandleUpdateSource(const etcpal::Uuid& cid, const std::string& name, const uint16_t* sourced_universes,
```

Comment:
For the C++ API, can we translate sourced_universes into a vector rather than pointer and length for nicer syntax?

Reply:
I'm on the fence --The main reason I didn't do this was that it would be copying memory, and if the library was set to static memory this would be a conflict.  This is why std::span would be so darn useful.  Given that this api will probably only be used in the dynamic memory case, I could easily be convinced.  Do you still want this?

Reply:
RDMnet uses allocating containers in the C++ API regardless of the library's memory configuration. I was just looking to copy that pattern. It's up to you.

Reply:
That's reasonable -- I've implemented that change.